langchain4j-agentic/src/main/java/dev/langchain4j/agentic/internal/PlannerBasedInvocationHandler.java

```diff
@@ -0,0 +1,297 @@
+
+    private final AbstractServiceBuilder<?, ?> service;
+
+    private final Supplier<Planner> plannerSupplier;
+
+    public PlannerBasedInvocationHandler(AbstractServiceBuilder<?, ?> service, Supplier<Planner> plannerSupplier) {
+        this(service, plannerSupplier, null);
+    }
+
+    private PlannerBasedInvocationHandler(AbstractServiceBuilder<?, ?> service, Supplier<Planner> plannerSupplier, DefaultAgenticScope agenticScope) {
+        this.service = service;
+        this.agentServiceClass = service.agentServiceClass;
+        this.output = service.output;
+        this.executor = service.executor;
+        this.agentInstances = service.agentExecutors().stream().map(AgentInstance.class::cast).toList();
+        this.name = service.name;
+        this.uniqueName = uniqueAgentName(this.name);
+        this.description = service.description;
+        this.outputKey = service.outputKey;
+        this.beforeCall = service.beforeCall;
+        this.errorHandler = service.errorHandler;
+        this.beforeListener = service.beforeListener;
+        this.afterListener = service.afterListener;
+        this.plannerSupplier = plannerSupplier;
+        this.agenticScope = agenticScope;
+        this.plannerInstance = AgentUtil.agentToExecutor(AgentUtil.buildAgent(agentServiceClass, this));
+    }
+
+    public AgenticScopeOwner withAgenticScope(DefaultAgenticScope agenticScope) {
+        return (AgenticScopeOwner) Proxy.newProxyInstance(
+                agentServiceClass.getClassLoader(),
+                new Class<?>[] {agentServiceClass, AgentSpecification.class, AgenticScopeOwner.class},
+                new PlannerBasedInvocationHandler(service, plannerSupplier, agenticScope));
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) {
+        AgenticScopeRegistry registry = agenticScopeRegistry();
+        if (method.getDeclaringClass() == AgenticScopeOwner.class) {
+            return switch (method.getName()) {
+                case "withAgenticScope" -> withAgenticScope((DefaultAgenticScope) args[0]);
+                case "registry" -> registry;
+                default ->
+                    throw new UnsupportedOperationException(
+                            "Unknown method on AgenticScopeOwner class : " + method.getName());
+            };
+        }
+
+        if (method.getDeclaringClass() == AgenticScopeAccess.class) {
+            return switch (method.getName()) {
+                case "getAgenticScope" -> registry.get(args[0]);
+                case "evictAgenticScope" -> registry.evict(args[0]);
+                default ->
+                    throw new UnsupportedOperationException(
+                            "Unknown method on AgenticScopeAccess class : " + method.getName());
+            };
+        }
+
+        if (method.getDeclaringClass() == AgentSpecification.class) {
+            return switch (method.getName()) {
+                case "name" -> name;
```

Comment:
Maybe use some String constants or enum to avoid typos in strings, to make it easier for later refactorings should they happen?